src/components/shared/thread/ActivityGraphCanvas.js

```diff
@@ -0,0 +1,224 @@
+  IndexIntoSamplesTable,
+  CategoryList,
+} from 'firefox-profiler/types';
+
+import type { CategoryDrawStyles } from './ActivityGraphFills';
+
+type CanvasProps = {|
+  +className: string,
+  +trackName: string,
+  +fullThread: Thread,
+  +interval: Milliseconds,
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +samplesSelectedStates: null | SelectedState[],
+  +treeOrderSampleComparator: (
+    IndexIntoSamplesTable,
+    IndexIntoSamplesTable
+  ) => number,
+  +categories: CategoryList,
+  +passFillsQuerier: any,
+  +onMouseUp: any,
+|};
+
+export class ActivityGraphCanvas extends React.PureComponent<CanvasProps> {
+  _canvas: {| current: null | HTMLCanvasElement |} = React.createRef();
+  _categoryDrawStyles: null | CategoryDrawStyles = null;
+
+  _renderCanvas() {
+    const canvas = this._canvas.current;
+    if (canvas !== null) {
+      timeCode('ThreadActivityGraph render', () => {
+        this.drawCanvas(canvas);
+      });
+    }
+  }
+
+  /**
+   * Get or lazily create the category info. It requires the 2d ctx to exist in order
+   * to create the fill patterns.
+   */
+  _getCategoryDrawStyles(ctx: CanvasRenderingContext2D): CategoryDrawStyles {
+    if (this._categoryDrawStyles === null) {
+      // Lazily initialize this list.
+      this._categoryDrawStyles = this.props.categories.map(
+        ({ color: colorName }, categoryIndex) => {
+          const styles = mapCategoryColorNameToStyles(colorName);
+          return {
+            ...styles,
+            category: categoryIndex,
+            filteredOutByTransformFillStyle: _createDiagonalStripePattern(
+              ctx,
+              styles.unselectedFillStyle
+            ),
+          };
+        }
+      );
+    }
+
+    return this._categoryDrawStyles;
+  }
+
+  componentDidMount() {
+    this._renderCanvas();
+  }
+
+  componentDidUpdate() {
+    this._renderCanvas();
+  }
+
+  drawCanvas(canvas: HTMLCanvasElement) {
+    const {
+      fullThread,
+      interval,
+      rangeStart,
+      rangeEnd,
+      samplesSelectedStates,
+      treeOrderSampleComparator,
+      categories,
+    } = this.props;
+
+    const rect = canvas.getBoundingClientRect();
+    const ctx = canvas.getContext('2d');
+    const canvasPixelWidth = Math.round(rect.width * window.devicePixelRatio);
+    const canvasPixelHeight = Math.round(rect.height * window.devicePixelRatio);
+    canvas.width = canvasPixelWidth;
+    canvas.height = canvasPixelHeight;
+
+    const { fills, fillsQuerier } = computeActivityGraphFills({
+      canvasPixelWidth,
+      canvasPixelHeight,
+      fullThread,
+      interval,
+      rangeStart,
+      rangeEnd,
+      samplesSelectedStates,
+      xPixelsPerMs: canvasPixelWidth / (rangeEnd - rangeStart),
+      treeOrderSampleComparator,
+      greyCategoryIndex: categories.findIndex(c => c.color === 'grey') || 0,
+      categoryDrawStyles: this._getCategoryDrawStyles(ctx),
+    });
+
+    this.props.passFillsQuerier(fillsQuerier);
```

Comment:
fillsQuerier is needed in ActivityGraph and is only computed in the drawCanvas method which has been extracted to ActivityGraphCanvas. So, I had to pass its value through the passFillsQuerier custom prop.

Reply:
This message would be good to leave as a comment in the source code.

Reply:
Done